### PR TITLE
Updates to likelihoods, bug fix `GaussianLogLikelihood` class

### LINKED ISSUE #257: [Bug]: GaussianLogLikelihood returns an error in the case of multiple signals
### Python Version

3.10.12

### Describe the bug

The `evaluate` and `evaluateS1` functions need to be updated for `GaussianLogLikelihood` to allow with multiple signals. Note that `GaussianLogLikelihoodKnownSigma` runs without error and produces the expected output.

### Steps to reproduce the behaviour

Replace lines 46-47 in the [spm_MLE.py](https://github.com/pybop-team/PyBOP/blob/develop/examples/scripts/spm_MLE.py) example:
```
problem = pybop.FittingProblem(model, parameters, dataset, signal=["Voltage [V]", "Time [s]"])
likelihood = pybop.GaussianLogLikelihoodKnownSigma(problem, sigma=[0.03, 0.03])
```
with:
```
problem = pybop.FittingProblem(model, parameters, dataset, signal=["Voltage [V]", "Time [s]"])
likelihood = pybop.GaussianLogLikelihood(problem)
```
Running this updated example produces the error message copied below.

### Relevant log output

```shell
----------------------------------------
Unexpected termination.
Current score: (-inf, -inf)
Current position:
 6.14848282459801054e-01
 4.47667093256804038e-01
----------------------------------------
Traceback (most recent call last):
  File "/home/nicola/GitHub/PyBOP/pybop/costs/base_cost.py", line 88, in evaluate
    return -self._evaluate(x, grad)
  File "/home/nicola/GitHub/PyBOP/pybop/costs/_likelihoods.py", line 148, in _evaluate
    y = self.problem.evaluate(x[: -self.n_outputs])
  File "/home/nicola/GitHub/PyBOP/pybop/_problem.py", line 244, in evaluate
    if np.any(x != self.x) and self._model.matched_parameters:
ValueError: operands could not be broadcast together with shapes (0,) (2,) 

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/nicola/GitHub/PyBOP/examples/scripts/spm_MLE.py", line 54, in <module>
    x, final_cost = optim.run()
  File "/home/nicola/GitHub/PyBOP/pybop/_optimisation.py", line 158, in run
    x, final_cost = self._run_pints()
  File "/home/nicola/GitHub/PyBOP/pybop/_optimisation.py", line 260, in _run_pints
    fs = evaluator.evaluate(xs)
  File "/home/nicola/GitHub/pybop-env/lib/python3.10/site-packages/pints/_evaluation.py", line 119, in evaluate
    return self._evaluate(positions)
  File "/home/nicola/GitHub/pybop-env/lib/python3.10/site-packages/pints/_evaluation.py", line 481, in _evaluate
    scores[k] = self._function(x, *self._args)
  File "/home/nicola/GitHub/PyBOP/pybop/costs/base_cost.py", line 60, in __call__
    return self.evaluate(x, grad)
  File "/home/nicola/GitHub/PyBOP/pybop/costs/base_cost.py", line 94, in evaluate
    raise ValueError(f"Error in cost calculation: {e}")
ValueError: Error in cost calculation: operands could not be broadcast together with shapes (0,) (2,)
```

### PR DESCRIPTION
**Pull request by BradyPlanden**
# Description

- Fixes `GaussianLogLikelihood` for use with all optimisers,
- Adds `GaussianLogLikelihood` to integration testing, 
- Updates non-bounded parameter logic, 
- Bugfix `CMAES` construction where `x0` is destroyed.
- Moves `MAP` to _likelihoods.py

## Issue reference
Fixes #257

## Review
Before you mark your PR as ready for review, please ensure that you've considered the following:
- Updated the [CHANGELOG.md](https://github.com/pybop-team/PyBOP/blob/develop/CHANGELOG.md) in reverse chronological order (newest at the top) with a concise description of the changes, including the PR number.
- Noted any breaking changes, including details on how it might impact existing functionality.

## Type of change
- [ ] New Feature: A non-breaking change that adds new functionality.
- [ ] Optimization: A code change that improves performance.
- [ ] Examples: A change to existing or additional examples.
- [x] Bug Fix: A non-breaking change that addresses an issue.
- [ ] Documentation: Updates to documentation or new documentation for new features.
- [x] Refactoring: Non-functional changes that improve the codebase.
- [ ] Style: Non-functional changes related to code style (formatting, naming, etc).
- [ ] Testing: Additional tests to improve coverage or confirm functionality.
- [ ] Other: (Insert description of change)

# Key checklist:

- [ ] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybop-team/PyBOP/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [ ] All unit tests pass: `$ nox -s tests`
- [ ] The documentation builds: `$ nox -s doctest`

You can run integration tests, unit tests, and doctests together at once, using `$ nox -s quick`.

## Further checks:
- [ ] Code is well-commented, especially in complex or unclear areas.
- [ ] Added tests that prove my fix is effective or that my feature works.
- [ ] Checked that coverage remains or improves, and added tests if necessary to maintain or increase coverage.

Thank you for contributing to our project! Your efforts help us to deliver great software.

### CODE REVIEW

File: examples/scripts/spm_MLE.py [hunk 1]
```diff
@@ -44,7 +43,7 @@
 
 # Generate problem, cost function, and optimisation class
 problem = pybop.FittingProblem(model, parameters, dataset)
-likelihood = pybop.GaussianLogLikelihoodKnownSigma(problem, sigma=[0.03, 0.03])
+likelihood = pybop.GaussianLogLikelihoodKnownSigma(problem, sigma0=0.005)
```
Comment (BradyPlanden) [inline examples/scripts/spm_MLE.py:46]:
```suggestion
likelihood = pybop.GaussianLogLikelihoodKnownSigma(problem, sigma0=sigma)
```

File: pybop/costs/_likelihoods.py [hunk 2]
```diff
@@ -27,54 +30,25 @@ class GaussianLogLikelihoodKnownSigma(BaseLikelihood):
         per dimension. Not all methods will use this information.
     """
 
-    def __init__(self, problem, sigma):
+    def __init__(self, problem: BaseProblem, sigma0: float):
```
Comment (BradyPlanden) [inline pybop/costs/_likelihoods.py:33]:
```suggestion
    def __init__(self, problem: BaseProblem, sigma0: Union[List[float], float]):
```

File: pybop/costs/_likelihoods.py [hunk 3]
```diff
@@ -84,29 +58,32 @@ def _evaluate(self, x, grad=None):
...
-
-        for key in self.signal:
-            if len(y.get(key, [])) != len(self._target.get(key, [])):
-                likelihood = np.float64(np.inf)
-                dl = self._dl * np.ones(self.n_parameters)
-                return -likelihood, -dl
+        if any(
+            len(y.get(key, [])) != len(self._target.get(key, [])) for key in self.signal
+        ):
+            return -np.inf, -self._dl * np.ones(self.n_parameters)
```
Comment (BradyPlanden) [inline pybop/costs/_likelihoods.py:71]:
```suggestion
            return -np.inf, -self._dl
```

File: pybop/costs/_likelihoods.py [hunk 4]
```diff
@@ -84,29 +58,32 @@ def _evaluate(self, x, grad=None):
...
 
         r = np.array([self._target[signal] - y[signal] for signal in self.signal])
         likelihood = self._evaluate(x)
         dl = np.sum((self.sigma2 * np.sum((r * dy.T), axis=2)), axis=1)
         return likelihood, dl
 
+    def check_sigma0(self, sigma0: Union[np.ndarray, float]):
+        """
+        Setter for sigma0 parameter
```
Comment (BradyPlanden) [inline pybop/costs/_likelihoods.py:80]:
```suggestion
        Check and set sigma0 variable.
```

File: pybop/costs/_likelihoods.py [hunk 5]
```diff
@@ -120,35 +97,81 @@ class GaussianLogLikelihood(BaseLikelihood):
         Precomputed offset value for the log-likelihood function.
     """
 
-    def __init__(self, problem):
+    def __init__(
+        self,
+        problem: BaseProblem,
+        sigma0=0.001,
```
Comment (BradyPlanden) [inline pybop/costs/_likelihoods.py:103]:
```suggestion
        sigma0=0.002,
```

File: pybop/costs/_likelihoods.py [hunk 6]
```diff
@@ -120,35 +97,81 @@ class GaussianLogLikelihood(BaseLikelihood):
...
+        x0=0.005,
+        sigma_bounds_std=6,
+        dsigma_scale=1,
+    ):
         super(GaussianLogLikelihood, self).__init__(problem)
         self._logpi = -0.5 * self.n_time_data * np.log(2 * np.pi)
-        self._dl = np.ones(self.n_parameters + self.n_outputs)
+        self._dl = np.inf * np.ones(self.n_parameters + self.n_outputs)
```
Comment (BradyPlanden) [inline pybop/costs/_likelihoods.py:110]:
```suggestion
        self._dl = np.ones(self.n_parameters + self.n_outputs)
```

File: pybop/costs/_likelihoods.py [hunk 7]
```diff
@@ -120,35 +97,81 @@ class GaussianLogLikelihood(BaseLikelihood):
...
         sigma = np.asarray(x[-self.n_outputs :])
-
         if np.any(sigma <= 0):
             return -np.inf
 
         y = self.problem.evaluate(x[: -self.n_outputs])
+        if any(
+            len(y.get(key, [])) != len(self._target.get(key, [])) for key in self.signal
+        ):
+            return -np.inf  # prediction doesn't match target
```
Comment (BradyPlanden) [inline pybop/costs/_likelihoods.py:172]:
```suggestion
            return -np.inf  # prediction length doesn't match target
```

File: pybop/costs/_likelihoods.py [hunk 8]
```diff
@@ -27,54 +30,25 @@ class GaussianLogLikelihoodKnownSigma(BaseLikelihood):
...
-        """
-        Calls the problem.evaluate method and calculates
-        the log-likelihood
+        Evaluates the Gaussian log-likelihood for the given parameters with known sigma.
         """
         y = self.problem.evaluate(x)
+        if any(
+            len(y.get(key, [])) != len(self._target.get(key, [])) for key in self.signal
+        ):
+            return -np.inf  # prediction doesn't match target
```
Comment (BradyPlanden) [inline pybop/costs/_likelihoods.py:49]:
```suggestion
            return -np.inf  # prediction length doesn't match target
```

File: tests/integration/test_optimisation_options.py [hunk 9]
```diff
@@ -67,7 +67,7 @@ def spm_costs(self, model, parameters, cost_class):
         # Define the cost to optimise
         problem = pybop.FittingProblem(model, parameters, dataset, init_soc=init_soc)
         if cost_class in [pybop.GaussianLogLikelihoodKnownSigma]:
-            return cost_class(problem, sigma=[0.03, 0.03])
+            return cost_class(problem, sigma0=[0.03, 0.03])
```
Comment (BradyPlanden) [inline tests/integration/test_optimisation_options.py:70]:
```suggestion
            return cost_class(problem, sigma0=0.002)
```

**Comment by NicolaCourtier** [comment]
Hi @BradyPlanden, thanks for working on this issue! I have tried adding `sigma` as a `Parameter` in #352 - please can you take a look and see what you think of this alternative implementation?

File: pybop/parameters/parameter.py [hunk 10]
```diff
@@ -149,9 +152,23 @@ def set_bounds(self, bounds=None):
             else:
                 self.lower_bound = bounds[0]
                 self.upper_bound = bounds[1]
+        elif self.prior is not None:
+            self.lower_bound = self.prior.mean - boundary_multiplier * self.prior.sigma
+            self.upper_bound = self.prior.mean + boundary_multiplier * self.prior.sigma
+            bounds = [self.lower_bound, self.upper_bound]
```
Comment (NicolaCourtier) [inline pybop/parameters/parameter.py:158]: Please add an output message here, to warn users that bounds have been applied - similar to the warning when bounds are not being applied.

File: CHANGELOG.md [hunk 11]
```diff
@@ -25,6 +25,7 @@
 
 ## Bug Fixes
 
+- [#338](https://github.com/pybop-team/PyBOP/pull/338) - Fixes GaussianLogLikelihood class, adds integration tests, updates non-bounded parameter implementation and bugfix to CMAES construction.
```
Comment (NicolaCourtier) [inline CHANGELOG.md:28]: Please clarify "updates non-bounded parameter implementation", perhaps: applies default bounds to parameters with priors?

**Review by NicolaCourtier — APPROVED** [review]
Great work, thanks @BradyPlanden! I left a few small comments but the implementation looks good overall.

**Pull request merged**